myia/prim/shape_inferrers.py

```diff
@@ -0,0 +1,157 @@
+"""Definition of shape inference for primitives."""
+
+import operator
+from functools import partial, reduce
+
+from ..infer import ANYTHING, GraphInferrer, register_inferrer, \
+    Track, MyiaShapeError, Inferrer
+from ..ir import Graph
+from ..dtype import Array
+
+from . import ops as P
+from .ops import Primitive
+
+
+def prod(iterable):
+    """Return the product of the elements of the iterator."""
+    return reduce(operator.mul, iterable, 1)
+
+
+shape_inferrer_constructors = {}
+
+
+class DefaultShapeInferrer(Inferrer):
+    """Shape inferrer for all primitves that don't take arrays."""
+
+    def __init__(self, engine):
+        """Initialize the DefaultShapeInferrer."""
+        super().__init__(engine, 'default_shape_inferrer')
+
+    async def __call__(self, *args):
+        """Since no arrays are involved, the shape is always ()."""
+        return ()
+
+    def provably_equivalent(self, other):
+        """This is always equal to itself."""
+        return type(self) == type(other)
+
+
+class ShapeTrack(Track):
+    """Infer the shape of a constant."""
+
+    def __init__(self, engine, name, *,
+                 constructors=shape_inferrer_constructors):
+        """Intialize a ShapeTrack."""
+        super().__init__(engine, name)
+        self.constructors = constructors
+
+    def from_value(self, v, context):
+        """Infer the shape of a constant."""
+        if isinstance(v, Primitive):
+            if v in self.constructors:
+                return self.constructors[v](self.engine)
+            else:
+                return DefaultShapeInferrer(self.engine)
+        elif isinstance(v, Graph):
+            return GraphInferrer(self.engine, 'shape', v, context)
+        else:
+            return getattr(v, 'shape', ())
+
+
+shape_inferrer = partial(register_inferrer,
+                         constructors=shape_inferrer_constructors)
+
+
+@shape_inferrer(P.return_, nargs=1)
+async def infer_shape_return(engine, v):
+    """Infer the shape of return."""
+    return await engine.get('shape', v)
+
+
+@shape_inferrer(P.if_, nargs=3)
+async def infer_shape_if(engine, cond, tb, fb):
+    """Infer the shape of if."""
+    tb_inf = await engine.get('shape', tb)
+    fb_inf = await engine.get('shape', fb)
+    v = await engine.get('value', cond)
+    if v is True:
+        # We only visit the first branch if the condition is provably true
+        return await tb_inf()
+    elif v is False:
+        # We only visit the second branch if the condition is provably false
+        return await fb_inf()
+    elif v is ANYTHING:
+        # The first branch to finish will return immediately. When the other
+        # branch finishes, its result will be checked against the other.
+        return await engine.assert_same('shape', tb_inf(), fb_inf())
```

Comment:
For safety's sake, I think there should be an else clause that raises an exception if v is somehow not True, False or ANYTHING (e.g. I'm pondering adding an ABSENT keyword for tracks that are not being inferred at all). Same goes for `infer_type_if` which has the same oversight.

Reply:
With the current code, if we aren't inferring values, you would get an exception.

Also, if it's not True, False, or ANYTHING, you'll get an error in the type checker (since the type will not be `Bool()`).

Reply:
I've added a safety branch.